Fix misleading Utility message.

phrohdoh opened the pull request:
Closes #9713.

atlimit8 commented:
I don't like identifying exception type by checking the message string.
A subclass should be used:
```c#
public class NoSuchCommandException : ArgumentException
{
	public NoSuchCommandException(string command)
		: base("No such command '{0}'", command) { }
}
```
```diff
@@ -57,7 +57,7 @@ static void Main(string[] args)
 			try
 			{
 				if (!actions.ContainsKey(args[0]))
-					throw new ArgumentException();
+					throw new NoSuchCommandException(args[0]);
 
 				var action = actions[args[0]].Key;
 				var validateActionArgs = actions[args[0]].Value;
@@ -78,8 +78,10 @@ static void Main(string[] args)
 				Log.Write("utility", "Received args: {0}", args.JoinWith(" "));
 				Log.Write("utility", "{0}", e);
 
-			    if (e is ArgumentException)
-			        Console.WriteLine("No such command '{0}'", args[0]);
+				if (e is NoSuchCommandException)
+				{
+					Console.WriteLine(e.Message);
+				}
 			    else
 			    {
                     Console.WriteLine("Error: Utility application crashed. See utility.log for details");
```

Phrohdoh commented:
That's fair.
I'll fix this soon, it is already 12:15am here.

phrohdoh committed:
Fix misleading Utility message

We were reporting that the entered command did not exist
by assuming that any ArgumentException indicated this.

Now we will throw and catch a NoSuchCommandException.

phrohdoh committed:
Cache the command entered in a variable in Utility.Program.cs for easier reading.

Phrohdoh commented:
Updated.

atlimit8 commented:
:+1: 
The `command` variable also makes the code a little more readable.

RoosterDragon commented on OpenRA.Utility/Program.cs:
```diff
@@ -15,6 +15,16 @@
 
 namespace OpenRA.Utility
 {
+	public class NoSuchCommandException : ArgumentException
```
This should inherit plain `Exception`.

atlimit8 replied:
How is `NoSuchCommandException` not an `ArgumentException` when it is about an argument.

RoosterDragon replied:
`ArgumentException` defines `ParamName` which hasn't been set at all here, so either that's an oversight or `ArgumentException` doesn't apply in the first place.

ScottNZ replied:
Furthermore, ArgumentExceptions are used for method arguments/parameters, not arguments for commands in OpenRA.Utility

atlimit8 replied:
OK.

Phrohdoh replied:
Fixed.

RoosterDragon commented on OpenRA.Utility/Program.cs:
```diff
@@ -78,13 +89,13 @@ static void Main(string[] args)
-                    throw;
-			    }
+				if (e is NoSuchCommandException)
+					Console.WriteLine(e.Message);
```
This cleanup enables you to `catch (NoSuchCommandException)` and do the error handling there, which would be better.

atlimit8 replied:
Then log output would have to be copied or placed in a method.
```c#
 				Log.Write("utility", "Received args: {0}", args.JoinWith(" "));
 				Log.Write("utility", "{0}", e);
```

Phrohdoh replied:
I'd rather keep the amount of changes introduced here to a minimum.

RoosterDragon replied:
Good points.

Phrohdoh commented:
`NoSuchCommandException` now inherits from `Exception`.

RoosterDragon commented:
:+1:

RoosterDragon commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/25ef0d4f364aa966dea233e24d7335df8630009f...3d276d29317d7ac22af2e4e3ece69bc3842f0ca7)
